File: app/src/main/java/com/example/review/MoveDataActivity.java
package com.example.review;

import android.content.Intent;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.review.Activity.MainActivity;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketAddress;
import java.util.ArrayList;

public class MoveDataActivity extends AppCompatActivity {

    EditText serviceId;
    EditText servicePort;
    Button   serviceBuild;
    EditText clientId;
    EditText clientPort;
    Button   clientSend;

    int port;
    private String TAG = "msg_mine";

    private static final int HANDLER_SET_PORT_ENABLE = 306;
    private static final int HANDLER_TOAST           = 307;

    Handler handler = new Handler(new Handler.Callback() {
        @Override
        public boolean handleMessage(Message message) {
            switch (message.what) {
                case HANDLER_SET_PORT_ENABLE:
                    if (message.arg1 == 0) {
                        servicePort.setEnabled(false);
                        serviceBuild.setEnabled(false);

                        clientId.setEnabled(false);
                        clientPort.setEnabled(false);
                        clientSend.setEnabled(false);
                        clientReceive.setEnabled(false);
                    } else {
                        servicePort.setEnabled(true);
                        serviceBuild.setEnabled(true);

                        clientId.setEnabled(true);
                        clientPort.setEnabled(true);
                        clientSend.setEnabled(true);
                        clientReceive.setEnabled(true);
                    }
                    break;
                case HANDLER_TOAST:
                    if (message.obj != null) {
                        Toast.makeText(MoveDataActivity.this, ((String) message.obj), Toast.LENGTH_SHORT).show();
                    } else {
                        if (message.arg1 == 1)
                            Toast.makeText(MoveDataActivity.this, "接收异常!", Toast.LENGTH_SHORT).show();
                        else
                            Toast.makeText(MoveDataActivity.this, "发送异常!", Toast.LENGTH_SHORT).show();
                    }
                    break;
            }
            return false;
        }
    });
    private TextView  tittle;
    private String    address;
    private Button    clientReceive;
    private String    order = "give me some data";
    private ImageView backButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_move_data);

        initViews();

        clientSend.setOnClickListener(clientSend());
        serviceBuild.setOnClickListener(serviceBuild());
        clientReceive.setOnClickListener(clientReceive());
        String localIpAddress = getLocalIpAddress();
        serviceId.setText(localIpAddress);
        serviceId.setEnabled(false);

        String name = String.format("同步到局域网的另一端？\n\n>> %s\n>> %s", MainActivity.pathNexus.getName(), MainActivity.pathLibrary.getName());
        tittle.setText(name);
        backButton.setOnClickListener(view -> finish());
    }

    private View.OnClickListener clientReceive() {
        return view -> {
            String        strAddress = clientId.getText().toString();
            int           port       = Integer.parseInt(clientPort.getText().toString());
            Socket        socket     = new Socket();
            SocketAddress address    = new InetSocketAddress(strAddress, port);

            new Thread(() -> {
                try {
                    socket.connect(address, 1000);
                    InputStream           in   = socket.getInputStream();
                    OutputStream          out  = socket.getOutputStream();
                    ByteArrayOutputStream baos = new ByteArrayOutputStream();

                    out.write(order.getBytes());
                    socket.shutdownOutput();

                    byte[] bytes = new byte[1024 * 1024];
                    int    length;

                    while ((length = in.read(bytes)) != -1) baos.write(bytes, 0, length);
                    byte[] buffer = baos.toByteArray();


                    ArrayList<String> paths = new ArrayList<>();
                    dataProcess(paths, buffer);

                    Message msg = new Message();
                    msg.what = HANDLER_TOAST;
                    msg.obj  = "success";
                    handler.sendMessage(msg);

                    in.close();
                    out.close();
                    socket.close();

                } catch (IOException e) {
                    e.printStackTrace();
                }
            }).start();
        };
    }

    private View.OnClickListener serviceBuild() {
        return view -> {
            //从界面读取数据
            port = Integer.parseInt(servicePort.getText().toString());

            new Thread(() -> {
                Message msg;
                try {
                    //发送Handler消息
                    msg      = new Message();
                    msg.what = HANDLER_SET_PORT_ENABLE;
                    msg.arg1 = 0;
                    handler.sendMessage(msg);

                    ArrayList<String> paths = service();
                    if (paths.size() > 0) {
                        Intent data = new Intent();
                        data.putStringArrayListExtra("paths", paths);
                        setResult(1, data);
                        finish();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                    msg      = new Message();
                    msg.what = HANDLER_TOAST;
                    msg.arg1 = 1;
                    handler.sendMessage(msg);
                } finally {
                    //发送Handler消息
                    msg      = new Message();
                    msg.what = HANDLER_SET_PORT_ENABLE;
                    msg.arg1 = 1;
                    handler.sendMessage(msg);
                }

            }).start();
        };
    }

    private ArrayList<String> service() throws IOException {
        ArrayList<String> paths = new ArrayList<>();

        //创建ServerSocket服务
        ServerSocket serverSocket = new ServerSocket(port);
        //等待连接
        Socket socket = serverSocket.accept();
        //连接成功，取输入流
        InputStream in = socket.getInputStream();


        //读取全部数据
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[]                temp = new byte[1024 * 1024];
        int                   index;

        //分段读取数据
        while ((index = in.read(temp)) != -1) baos.write(temp, 0, index);
        byte[] data = baos.toByteArray();


        if (data.length <= order.length()) {
            OutputStream out = socket.getOutputStream();

            File            pathNexus = MainActivity.pathNexus;
            FileInputStream fis       = new FileInputStream(pathNexus);
            long            length    = pathNexus.length();
            byte[]          buffer    = new byte[(int) length];
            fis.read(buffer);
            fis.close();
            baos.reset();
            DataOutputStream dos = new DataOutputStream(baos);

            dataStream(dos, pathNexus.getName().getBytes());
            dataStream(dos, buffer);

            File pathLibrary = MainActivity.pathLibrary;
            fis    = new FileInputStream(pathLibrary);
            length = pathLibrary.length();
            buffer = new byte[(int) length];
            fis.read(buffer);
            fis.close();

            dataStream(dos, pathLibrary.getName().getBytes());
            dataStream(dos, buffer);
            data = baos.toByteArray();
            dos.close();

            out.write(data);
            socket.shutdownOutput();

            //关闭连接、流
            out.close();
            in.close();
            socket.close();
            serverSocket.close();

        } else {
            //关闭连接、流
            socket.close();
            serverSocket.close();

            dataProcess(paths, data);
        }
        return paths;
    }

    private void dataProcess(ArrayList<String> paths, byte[] data) throws IOException {
        //包装类
        ByteArrayInputStream bais = new ByteArrayInputStream(data);
        DataInputStream      dis  = new DataInputStream(bais);
        byte[]               buffer;

        for (int i = 0; i < 2; i++) {
            buffer = dataStream(dis);//读取文件名
            String name = new String(buffer);
            buffer = dataStream(dis);//读取数据


            //保存文件的路径
            File path = new File(MainActivity.pathApp, name);//todo
            paths.add(path.getPath());

            if (path.exists()) {
                //备份文件路径
                File pathBackup = new File(MainActivity.pathApp, name.concat(".backup"));
                //备份文件存在，则删除它
                if (pathBackup.exists()) pathBackup.delete();
                //把旧文件备份
                path.renameTo(pathBackup);
            }

            FileOutputStream out = new FileOutputStream(path);
            out.write(buffer);
            out.close();
        }
    }


    private View.OnClickListener clientSend() {
        return view -> new Thread(() -> client()).start();
    }

    private void client() {
        //从界面读取数据
        port    = Integer.valueOf(clientPort.getText().toString());
        address = clientId.getText().toString();

//        port    = Integer.valueOf(servicePort.getText().toString());
//        address = serviceId.getText().toString();

        String nameNex = MainActivity.pathNexus.getName();
        String nameLib = MainActivity.pathLibrary.getName();

        byte[] bytesNex = new byte[(int) MainActivity.pathNexus.length()];
        byte[] bytesLib = new byte[(int) MainActivity.pathLibrary.length()];

        FileInputStream fis;
        try {
            fis = new FileInputStream(MainActivity.pathNexus);
            fis.read(bytesNex);
            fis = new FileInputStream(MainActivity.pathLibrary);
            fis.read(bytesLib);
        } catch (IOException e) {
            e.printStackTrace();
        }

        try {
            //包装类
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            DataOutputStream      dos  = new DataOutputStream(baos);


            //数据通过dos汇集到baos中
            dataStream(dos, nameNex.getBytes());
            dataStream(dos, bytesNex);
            dataStream(dos, nameLib.getBytes());
            dataStream(dos, bytesLib);

            byte[]               buf  = baos.toByteArray();
            ByteArrayInputStream bais = new ByteArrayInputStream(buf);

            Socket       socket = new Socket(address, port);//客户端socket
            OutputStream out    = socket.getOutputStream();//取输出流


            //分段发送数据
            int    index;
            byte[] temp = new byte[1024 * 1024];
            while ((index = bais.read(temp)) != -1) out.write(temp, 0, index);
            socket.shutdownOutput();

            //关闭流
            dos.close();
            out.close();
            socket.close();

        } catch (IOException e) {
            Message msg = new Message();
            msg.what = HANDLER_TOAST;
            msg.arg1 = 2;
            handler.sendMessage(msg);
            e.printStackTrace();
        }
    }

    void dataStream(DataOutputStream dos, byte[] bytes) throws IOException {
        dos.writeInt(bytes.length);
        dos.write(bytes);
        Log.d(TAG, "dataStream out: size = " + bytes.length);
    }

    byte[] dataStream(DataInputStream dis) throws IOException {
        int    len   = dis.readInt();
        byte[] bytes = new byte[len];
        dis.read(bytes);
        Log.d(TAG, "dataStream in: size = " + bytes.length);
        return bytes;
    }

    private void initViews() {
        serviceId    = findViewById(R.id.moveData_et_ServiceIp);
        servicePort  = findViewById(R.id.moveData_et_ServicePort);
        serviceBuild = findViewById(R.id.moveData_et_ServiceBuild);

        clientId      = findViewById(R.id.moveData_et_ClientIp);
        clientPort    = findViewById(R.id.moveData_et_clientPort);
        clientSend    = findViewById(R.id.moveData_et_ClientSend);
        clientReceive = findViewById(R.id.moveData_et_ClientReceive);
        tittle        = findViewById(R.id.moveData_tv_tittle);
        backButton    = findViewById(R.id.moveData_imageView_back_button);
    }

    String getLocalIpAddress() {
        WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(WIFI_SERVICE);
        int         ipAddress   = 0;

        int wifiState = wifiManager.getWifiState();
        if (wifiState == WifiManager.WIFI_STATE_ENABLED) {
            WifiInfo connectionInfo = wifiManager.getConnectionInfo();
            ipAddress = connectionInfo.getIpAddress();
        }
        return int2ip(ipAddress);
    }

    String int2ip(int ipInt) {
        StringBuilder sb = new StringBuilder();
        sb.append(ipInt & 0xFF).append(".");
        sb.append((ipInt >> 8) & 0xFF).append(".");
        sb.append((ipInt >> 16) & 0xFF).append(".");
        sb.append((ipInt >> 24) & 0xFF);
        return sb.toString();
    }
}

File: app/src/main/java/com/example/review/Activity/MoveDataActivity.kt
package com.example.review.Activity

import android.content.Context
import android.content.Intent
import android.net.wifi.WifiManager
import android.os.Bundle
import android.os.Handler
import android.os.Handler.Callback
import android.os.Message
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View.OnClickListener
import android.widget.*
import com.example.review.R
import java.io.*
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketAddress

class MoveDataActivity : AppCompatActivity() {
    private lateinit var serviceId: EditText
    private lateinit var servicePort: EditText
    private lateinit var serviceBuild: Button
    private lateinit var clientId: EditText
    private lateinit var clientPort: EditText
    private lateinit var clientSend: Button
    private var port = 0
    private val TAG = "msg_mine"
    private var handler = Handler(Callback { message ->
        when (message.what) {
            HANDLER_SET_PORT_ENABLE -> setViewsEnabled(message.arg1 != 0)
            HANDLER_TOAST -> {
                val value = Toast.LENGTH_LONG
                if (message.obj != null)
                    Toast.makeText(this@MoveDataActivity, message.obj as String, value).show()
                else {
                    if (message.arg1 == 1) Toast.makeText(this@MoveDataActivity, "接收异常!", value).show()
                    else Toast.makeText(this@MoveDataActivity, "发送异常!", value).show()
                }
            }
        }
        false
    })
    private lateinit var tittle: TextView
    private lateinit var clientReceive: Button
    private lateinit var backButton: ImageView
    private var address: String? = null
    private val order = "give me some data"

    //设置一些特定的View的enable属性
    private fun setViewsEnabled(isEnable: Boolean = true) {
        servicePort.isEnabled = isEnable
        serviceBuild.isEnabled = isEnable
        clientId.isEnabled = isEnable
        clientPort.isEnabled = isEnable
        clientSend.isEnabled = isEnable
        clientReceive.isEnabled = isEnable
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_move_data)

        initViews()

        clientSend.setOnClickListener(clientSend())
        serviceBuild.setOnClickListener(serviceBuild())
        clientReceive.setOnClickListener(clientReceive())
        backButton.setOnClickListener { finish() }

        val localIpAddress = localIpAddress
        serviceId.setText(localIpAddress)
        serviceId.isEnabled = false
        tittle.text = String.format("同步到局域网的另一端？\n\n>> %s\n>> %s", MainActivity.pathNexus.name, MainActivity.pathLibrary.name)
    }

    /**
     * 接收服务端的数据
     * */
    private fun clientReceive(): OnClickListener {
        return OnClickListener {
            val strAddress = clientId.text.toString()
            val port = clientPort.text.toString().toInt()
            val socket = Socket()
            val address: SocketAddress = InetSocketAddress(strAddress, port)

            Thread(Runnable {
                try {
                    socket.connect(address, 1000)
                    val input = socket.getInputStream()
                    val output = socket.getOutputStream()
                    val baos = ByteArrayOutputStream()

                    output.write(order.toByteArray())
                    socket.shutdownOutput()

                    val bytes = ByteArray(1024 * 1024)
                    var length: Int
                    while (input.read(bytes).also { length = it } != -1) baos.write(bytes, 0, length)

                    val buffer = baos.toByteArray()
                    val paths = ArrayList<String>()
                    dataProcess(paths, buffer)

                    val msg = Message()
                    msg.what = HANDLER_TOAST
                    msg.obj = "接收成功！"
                    handler.sendMessage(msg)

                    if (paths.isNotEmpty()) {
                        val data = Intent()
                        data.putStringArrayListExtra("paths", paths)
                        setResult(1, data)
                        finish()
                    }

                    input.close()
                    output.close()
                    socket.close()
                } catch (e: IOException) {
                    val msg = Message()
                    msg.what = HANDLER_TOAST
                    msg.obj = "出现异常！\n$e"
                    handler.sendMessage(msg)
                }
            }).start()
        }
    }

    private fun serviceBuild(): OnClickListener {
        return OnClickListener {
            //从界面读取数据
            port = servicePort.text.toString().toInt()

            Thread(Runnable {
                var msg: Message
                try { //发送Handler消息
                    msg = Message()
                    msg.what = HANDLER_SET_PORT_ENABLE
                    msg.arg1 = 0
                    handler.sendMessage(msg)
                    val paths = service()

                    if (paths.size > 0) {
                        val data = Intent()
                        data.putStringArrayListExtra("paths", paths)
                        setResult(1, data)
                        finish()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                    msg = Message()
                    msg.what = HANDLER_TOAST
                    msg.arg1 = 1
                    handler.sendMessage(msg)
                } finally { //发送Handler消息
                    msg = Message()
                    msg.what = HANDLER_SET_PORT_ENABLE
                    msg.arg1 = 1
                    handler.sendMessage(msg)
                }
            }).start()
        }
    }

    @Throws(IOException::class)
    private fun service(): ArrayList<String> {
        val paths = ArrayList<String>()
        val serverSocket = ServerSocket(port)//创建ServerSocket服务
        val socket = serverSocket.accept()//等待连接
        val input = socket.getInputStream()//连接成功，取输入流
        val baos = ByteArrayOutputStream()//读取全部数据
        val temp = ByteArray(1024 * 1024)
        var index: Int

        //分段读取数据
        while (input.read(temp).also { index = it } != -1) baos.write(temp, 0, index)
        var data = baos.toByteArray()

        if (data.size <= order.length) {
            val out = socket.getOutputStream()
            val pathNexus: File = MainActivity.pathNexus
            var fis = FileInputStream(pathNexus)
            var length = pathNexus.length()
            var buffer = ByteArray(length.toInt())

            fis.read(buffer)
            fis.close()
            baos.reset()

            val dos = DataOutputStream(baos)
            dataStream(dos, pathNexus.name.toByteArray())
            dataStream(dos, buffer)

            val pathLibrary: File = MainActivity.pathLibrary
            fis = FileInputStream(pathLibrary)
            length = pathLibrary.length()
            buffer = ByteArray(length.toInt())
            fis.read(buffer)
            fis.close()

            dataStream(dos, pathLibrary.name.toByteArray())
            dataStream(dos, buffer)
            data = baos.toByteArray()
            dos.close()
            out.write(data)
            socket.shutdownOutput()

            //关闭连接、流
            out.close()
            input.close()
            socket.close()
            serverSocket.close()

        } else { //关闭连接、流
            socket.close()
            serverSocket.close()
            dataProcess(paths, data)
        }
        return paths
    }

    @Throws(IOException::class)
    private fun dataProcess(paths: ArrayList<String>, data: ByteArray) { //包装类
        val bais = ByteArrayInputStream(data)
        val dis = DataInputStream(bais)
        var buffer: ByteArray

        for (i in 0..1) {
            buffer = dataStream(dis) //读取文件名
            val name = String(buffer)
            buffer = dataStream(dis) //读取数据

            //保存文件的路径
            val path = File(MainActivity.pathApp, name) //todo
            paths.add(path.path)

            if (path.exists()) { //备份文件路径
                val pathBackup = File(MainActivity.pathApp, "$name.backup")
                if (pathBackup.exists()) pathBackup.delete()//备份文件存在，则删除它
                path.renameTo(pathBackup)//把旧文件备份
            } else {//文件不存在，则创建文件
                path.mkdirs()
                path.createNewFile()
            }

            val out = FileOutputStream(path)
            out.write(buffer)
            out.close()
        }
    }

    private fun clientSend(): OnClickListener {
        return OnClickListener { Thread(Runnable { client() }).start() }
    }

    private fun client() { //从界面读取数据
        port = Integer.valueOf(clientPort.text.toString())
        address = clientId.text.toString()
//        port    = Integer.valueOf(servicePort.getText().toString());
//        address = serviceId.getText().toString();
        val nameNex: String = MainActivity.pathNexus.name
        val nameLib: String = MainActivity.pathLibrary.name
        val bytesNex = ByteArray(MainActivity.pathNexus.length().toInt())
        val bytesLib = ByteArray(MainActivity.pathLibrary.length().toInt())
        var fis: FileInputStream
        try {
            fis = FileInputStream(MainActivity.pathNexus)
            fis.read(bytesNex)
            fis = FileInputStream(MainActivity.pathLibrary)
            fis.read(bytesLib)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        try { //包装类
            val baos = ByteArrayOutputStream()
            val dos = DataOutputStream(baos)

            //数据通过dos汇集到baos中
            dataStream(dos, nameNex.toByteArray())
            dataStream(dos, bytesNex)
            dataStream(dos, nameLib.toByteArray())
            dataStream(dos, bytesLib)

            val buf = baos.toByteArray()
            val bais = ByteArrayInputStream(buf)
            val socket = Socket(address, port) //客户端socket
            val out = socket.getOutputStream() //取输出流

            //分段发送数据
            var index: Int
            val temp = ByteArray(1024 * 1024)
            while (bais.read(temp).also { index = it } != -1) out.write(temp, 0, index)
            socket.shutdownOutput()

            //关闭流
            dos.close()
            out.close()
            socket.close()
        } catch (e: IOException) {
            val msg = Message()
            msg.what = HANDLER_TOAST
            msg.arg1 = 2
            handler.sendMessage(msg)
            e.printStackTrace()
        }
    }

    @Throws(IOException::class)
    internal fun dataStream(dos: DataOutputStream, bytes: ByteArray) {
        dos.writeInt(bytes.size)
        dos.write(bytes)
        Log.d(TAG, "dataStream out: size = " + bytes.size)
    }

    @Throws(IOException::class)
    internal fun dataStream(dis: DataInputStream): ByteArray {
        val len = dis.readInt()
        val bytes = ByteArray(len)
        dis.read(bytes)
        Log.d(TAG, "dataStream in: size = " + bytes.size)
        return bytes
    }

    private fun initViews() {
        serviceId = findViewById(R.id.moveData_et_ServiceIp)
        servicePort = findViewById(R.id.moveData_et_ServicePort)
        serviceBuild = findViewById(R.id.moveData_et_ServiceBuild)
        clientId = findViewById(R.id.moveData_et_ClientIp)
        clientPort = findViewById(R.id.moveData_et_clientPort)
        clientSend = findViewById(R.id.moveData_et_ClientSend)
        clientReceive = findViewById(R.id.moveData_et_ClientReceive)
        tittle = findViewById(R.id.moveData_tv_tittle)
        backButton = findViewById(R.id.moveData_imageView_back_button)
    }

    private val localIpAddress: String
        get() {
            val wifiManager = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            var ipAddress = 0
            val wifiState = wifiManager.wifiState

            if (wifiState == WifiManager.WIFI_STATE_ENABLED) {
                val connectionInfo = wifiManager.connectionInfo
                ipAddress = connectionInfo.ipAddress
            }
            return int2ip(ipAddress)
        }

    private fun int2ip(ipInt: Int): String {
        val sb = StringBuilder()
        sb.append(ipInt and 0xFF).append(".")
        sb.append(ipInt shr 8 and 0xFF).append(".")
        sb.append(ipInt shr 16 and 0xFF).append(".")
        sb.append(ipInt shr 24 and 0xFF)
        return sb.toString()
    }

    companion object {
        private const val HANDLER_SET_PORT_ENABLE = 306
        private const val HANDLER_TOAST = 307
    }
}